File: C++/Jeremy Game/Vec2.cpp
/* Start Header -------------------------------------------------------
File Name: Vector.cpp
Purpose: Vector
Language: C++, Visual Studio 2005
Platform: Visual Studio 8, Windows
Author: Jeremy Anderson
Creation date: Monday, January 20, 2009
- End Header --------------------------------------------------------*/
#include "Vec2.h"
#include <cmath>

Vec2::Vec2( const float x, const float y ) : x( x ), y( y )
{
}

Vec2::Vec2( float *rhs ) : x( rhs[0] ), y( rhs[1] )
{
}

Vec2 &Vec2::operator =( const Vec2 &rhs )
{
  //if ( rhs.w == 0.0f || rhs.w == 1.0f )
  //{
  x = rhs.x; y = rhs.y;
  //}
  //else
  //{
  //  x = rhs.x / rhs.w; y = rhs.y / rhs.w; z = rhs.z / rhs.w; w = 1.0f;
  //}

  return *this;
}


//Vec2::operator float *()
//{
//  return Data;
//}

Vec2 Vec2::operator -() const
{
  return Vec2( -x, -y );
}

Vec2 Vec2::operator +( const Vec2 &rhs ) const
{
  return Vec2( *this ) += rhs;
}

Vec2 Vec2::operator +( const float rhs ) const
{
  return Vec2( x + rhs, y + rhs );
}

Vec2 Vec2::operator -( const Vec2 &rhs ) const
{
  return Vec2( *this ) -= rhs;
}

Vec2 &Vec2::operator +=( const Vec2 &rhs )
{
  x += rhs.x; y += rhs.y;
  return *this; 
}

Vec2 &Vec2::operator -=( const Vec2 &rhs )
{
  x -= rhs.x; y -= rhs.y;
  return *this; 
}

Vec2 &Vec2::operator /=( const float rhs )
{
  x /= rhs; y /= rhs;
  return *this; 
}

Vec2 Vec2::operator /( const float rhs ) const
{
  Vec2 temp = *this;
  temp /= rhs;
  return temp; 
}

float Vec2::operator *( const Vec2 &rhs ) const
{
  return x * rhs.x + y * rhs.y;
}

Vec2 Vec2::operator *( const float s ) const
{
  return Vec2( x * s, y * s );
}


Vec2 Vec2::Normalize() const
{
  return *this / length();
}

float Vec2::dot( const Vec2& rhs ) const
{
  return x * rhs.x + y * rhs.y;
}

float Vec2::length( void ) const
{
  return sqrt( lengthSquared() );
}

float Vec2::lengthSquared( void ) const
{
  return this->dot( *this );
}

float Distance( const Vec2 &lhs, const Vec2 &rhs )
{
  Vec2 temp = lhs - rhs;
  return temp.length();
}

float DistanceSquared( const Vec2 &lhs, const Vec2 &rhs )
{
  Vec2 temp = lhs - rhs;
  return temp.lengthSquared();
}

File: C++/Jeremy Game/resources/scripts/vec2.lua
--- vec2
---
--- simple 2d vector class
--- Author - Jeremy Anderson



Vec2 = class()

function Vec2:init( x, y )
    if x == nil then
        self.x = 0.0
    else
        self.x = x
    end

    if y == nil then
        self.y = 0.0
    else
        self.y = y
    end
end

function Vec2:__add(rhs)
    return Vec2( self.x + rhs.x, self.y + rhs.y )
end

function Vec2:__sub(rhs)
    return Vec2( self.x - rhs.x, self.y - rhs.y )
end

--scale by
function Vec2:__mul(f)
    return Vec2( self.x * f, self.y * f )
end

--scale by
function Vec2:__div(f)
    return Vec2( self.x / f, self.y / f )
end

function Vec2:__eq(rhs)
    return self.x == rhs.x and self.y == rhs.y
end

function Vec2:__nq(rhs)
    return self.x ~= rhs.x and self.y ~= rhs.y
end

function Vec2:dot(rhs)
    return self.x * rhs.x + self.y * rhs.y
end

function Vec2:__unm()
    return Vec2( -self.x, -self.y )
end

--returns Vec2 as a table
function Vec2:get()
    return {self.x, self.y}
end

function Vec2:normalize()
    return self / self:length()
end

function Vec2:length()
    return math.sqrt( self:dot(self) )
end

function Vec2:lengthSquared()
    return self:dot(self)
end

function Vec2:printVars()
    print( self.x..' '..self.y )
end

function distance( lhs, rhs )
    temp = lhs - rhs
    return temp:length()
end

function distanceSquared( lhs, rhs )
    temp = lhs - rhs
    return temp:lengthSquared()
end

--a = -Vec2( 1.0, 2.0 )
--b = Vec2( 1.0, -1.0 )
--c = a + b
--d = a * 3.0
--e = Vec2( 1.0, 1.0 )

--a:printVars()
--b:printVars()
--c:printVars()
--d:printVars()
--print(distance(a,b))
